bot/bot.py

```diff
@@ -39,6 +45,69 @@ def __init__(self, redis_session: RedisSession, **kwargs):
+        # for that specific game leaderboard
+        self.games_leaderboard: dict[str, tuple[RedisCache, ...]] = dict()
+        self.auto_per_day_leaderboard_clear.start()
+
+    @loop()
+    async def auto_per_day_leaderboard_clear(self) -> None:
+        """Loop for clearing the per day leaderboard redis cache at UTC midnight."""
+        # once d.py get support for `time` parameter in loop decorator,
+        # this can be removed and the loop can use the `time=datetime.time.min` parameter
+        now = datetime.utcnow()
+        tomorrow = now + timedelta(days=1)
+        midnight_tomorrow = tomorrow.replace(hour=0, minute=0, second=0)
+
+        await sleep_until(midnight_tomorrow)
+
+        for _, per_day_lb in self.games_leaderboard.values():
+            await per_day_lb.clear()
+
+    def ensure_leaderboard(self, cog: Cog) -> None:
+        """Ensure that per day leaderboard and total leaderboard redis cache are available for the cog."""
+        if self.games_leaderboard.get(cog.qualified_name):
+            return
+
+        # This cache contains the leaderboard for game `cog`
+        # RedisCache[int: int]
+        # Where the key is the discord user ID, and the values if the total
+        # points registered for that user.
+        _leaderboard_cache = RedisCache(namespace=f"{cog.qualified_name}-total")
+
+        # This cache contains the per day leaderboard, same specifies
+        _per_day_leaderboard = RedisCache(namespace=cog.qualified_name)
+
+        self.games_leaderboard[cog.qualified_name] = (_leaderboard_cache, _per_day_leaderboard)
+
+    async def increment_points(self, cog: Cog, user: discord.Member, points: int = DEFAULT_POINTS) -> None:
+        """
+        Increment points for a user in the leaderboard for game `cog` by `points`.
+
+        If max points for today are hit for that particular game cog, then set today's points to the max
+        points and add the difference of the(before today points and max possible points for a day to the
+        overall points for the user.
+        """
+        _leaderboard_cache, _per_day_leaderboard = self.games_leaderboard.get(cog.qualified_name)
+
+        current_points = await _leaderboard_cache.get(user.id) or 0
+        current_points_today = await _per_day_leaderboard.get(user.id) or 0
+        new_points_today = current_points_today + points
+
+        if new_points_today > MAX_PER_GAME_PER_DAY_POINTS:
+            log.info(f"Member({user.id}) has got maximum possible points for game cog {cog.qualified_name}.")
+            await _per_day_leaderboard.set(user.id, MAX_PER_GAME_PER_DAY_POINTS)
+            await _leaderboard_cache.set(
+                user.id, current_points + (MAX_PER_GAME_PER_DAY_POINTS - current_points_today)
+            )
+        else:
+            await _per_day_leaderboard.set(user.id, new_points_today)
+            await _leaderboard_cache.set(user.id, current_points + points)
+            log.info(
```

Comment:
Maybe this one could be a DEBUG? It will most likely be called a lot.

Reply:
Debug logs aren't shown in the production setup, hence I kept it as `info`.